File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    candidates
    |> Enum.filter(fn(word) -> anagram?(base, word) end)
  end

  defp anagram?(base, word) do
    a = normalized(base)
    b = normalized(word)
    a != b and Enum.sort(a) == Enum.sort(b)
  end

  defp normalized(word) do
    word
    |> String.downcase
    |> String.to_char_list
  end
end

File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word
    @normalized = normalize @word
    @sorted = sort @normalized
  end

  def match(candidates)
    candidates.select { |w| anagram?(w) }
  end

  private

  def anagram?(candidate)
    candidate = normalize candidate
    @normalized != candidate && @sorted == sort(candidate)
  end

  def normalize(word)
    word.downcase
  end

  def sort(word)
    word.chars.sort
  end
end

module BookKeeping
  VERSION = 2
end
